Fix some minor issues of the DRBG reseeding

mspncp opened the pull request:
This PR contains three commits which are explained in detail separately. They are not intended to be squashed.

# Clarify difference between entropy counts and buffer lengths  (7ff7ddb653e4e50ae82e99bed9daff8f95cf2bf2)

Unlike the NIST DRBG standard, in OpenSSL the entropy counts are in bits and the buffer lengths are in bytes. The mixing of bits and bytes has occasionally lead to some confusion in the past, in fact to an error in the FIPS DRBG code (originally [RT 4055](https://rt.openssl.org/Ticket/Display.html?id=4055&user=guest&pass=guest), now Issue #2443) see [my comment added to crypto/rand/drbg_lib.c in PR #3789](https://github.com/openssl/openssl/pull/3789#pullrequestreview-48756739) (you have to click "Show outdated" to see it). To clarify the distinction between entropy counts and buffer lengths, it is useful to add a `len` suffix to all member names of `RAND_DRBG` which represent buffer lengths:
```
    -   {min,max}_{entropy,adin,nonce,pers}
    +   {min,max}_{entropy,adin,nonce,pers}len
```
This change also makes naming more consistent with existing names, as can be seen in diffs, for example:
```diff    
    -    else if (adinlen > drbg->max_adin) {
    +    else if (adinlen > drbg->max_adinlen) {
```

# Rename the RAND_poll_ex() callback and its typedef (0b2a15a6b7d6243032f8e9a2fa472c71867c74bd)

- [X] documentation is added or updated


With the introduction of RAND_poll_ex(), the good old `RAND_add()` calls were replaced by meaningless `cb(...)` calls. This commit changes the signature of `RAND_poll_ex()` from
```C
int RAND_poll_ex(RAND_poll_fn cb, void *arg);
```
to
```C
int RAND_poll_ex(RAND_add_ex_fn rand_add, void *arg);
```
such that all the nameless `cb(...)` callbacks are turned back into `rand_add(...)` calls.  Also, the misleading type name `RAND_poll_fn` is changed to `RAND_add_ex_fn`, since it's actually more a `RAND_add()` replacement. 

Actually, the `RAND_add_ex_fn` could have been called `RAND_add_fn`, if it weren't for the extra `void *arg` which is passed through, but used nowhere. Is this `arg` really needed? Of which use case are you thinking? Shouldn't it be removed again?


# Remove `randomness` buffer from `RAND_DRBG`  (8931a3741b066f27811fef9340e33645c5bf7d83)

In the original OpenSSL FIPS DRBG implementation, the function signature of the `get_entropy()` and `cleanup_entropy()` callbacks (resp. `get_nonce()` and `cleanup_nonce()`) is designed in such a way that the randomness buffer does not have to be allocated by the calling function: Instead, the calling function can pass the address of a stack local `char *randomness` pointer as `**pout` argument to the `get_entropy()` callback, which allocates the buffer and returns its address via `*pout`. After use, `address` is passed by the calling function to the `cleanup_entropy()` callback as `pout` parameter, where it is used to free the buffer.

```C
int FIPS_drbg_set_callbacks(
    DRBG_CTX *dctx,
    size_t (*get_entropy)(DRBG_CTX *ctx, unsigned
        char **pout, int entropy, size_t min_len, size_t max_len),
    void (*cleanup_entropy)(DRBG_CTX *ctx, unsigned char *out, size_t olen),
    
    ...
)
```

This implies that the pointer can be a stack local variable of the calling function, provided the two callbacks are paired properly. In particular, it is not necessary have a member in the `RAND_DRBG` struct. I pointed this out to Rich Salz a while ago and my suggestion was partly followed in commit 9d951a7872e5fa2b2a83fe8cfda3af5c52581172 which reduced the lifetime of the `randomness` buffer to the scope of the `RAND_DRBG_instantiate()` resp. `RAND_DRBG_reseed()` function. But this commit also removed the necessity to store the pointer in the `RAND_DRBG` struct, so I finally removed it.

mspncp commented:
CLA has been sent long time ago.

richsalz commented:
yes, i think we have a problem with case sensitivity :(

FdaSilvaYY commented on crypto/rand/rand_lib.c:
```diff
@@ -150,33 +143,31 @@ size_t drbg_entropy_from_parent(RAND_DRBG *drbg,
+    st = RAND_DRBG_generate(drbg->parent, randomness, min_len, 0,
                             (unsigned char *)drbg, sizeof(*drbg));
     if (st == 0)
         return 0;
```
It leaks on failure of  RAND_DRBG_generate, no ?

mspncp replied:
You're right, thanks! That's a consequence of the changed lifetime, which I overlooked.  I will fix this.

mspncp commented:
@richsalz: There is a long comment about bits & bytes in the commit message, have you seen it? But if you prefer, I can find an appropriate location for a comment, too.

mspncp commented:
@richsalz quoting from above:  Actually, the RAND_add_ex_fn could have been called RAND_add_fn, if it weren't for the extra void *arg which is passed through, but used nowhere. Is this arg really needed? Of which use case are you thinking? Shouldn't it be removed again?

richsalz commented:
Yes, subset the commit comment and put it into the source where folks will find it when the read the code; nobody reads commit logs :)

The void* argument is so you could call RAND_poll_ex and give a particular DRBG object as the void* to a generic "add sys entropy to this drbg" callback.   Yes that code's not written yet.

I don't like changing the typedef, since the callback is used by RAND_poll_ex.  It doesn't have to call RAND_add.  Changing the param name, doesn't matter as much.

mspncp commented:
I don't like the original name, because to me `RAND_poll_fn` suggests that the function pointer typedef has the same signature as `RAND_poll()`, which isn't the case. Since it's a callback for `RAND_poll()`, would you agree with changing the typedef name to `RAND_poll_cb_fn`?

richsalz commented:
We use the "cb" suffix elsewhere, so how about ``RAND_poll_cb``? :)

mspncp commented:
Just force pushed new commits, addressing all outstanding change requests by amending the appropriate commit:

* changed typedef name to `RAND_poll_cb`
* added bits&bytes comment inside definition of `struct rand_drbg_st`
* fixed leak on failure

Ready for final review. Please don't squash :-).

mspncp commented on crypto/rand/rand_lcl.h:
```diff
@@ -94,21 +94,34 @@ struct rand_drbg_st {
-    unsigned char *randomness;
 
-    /* These parameters are setup by the per-type "init" function. */
+    /* The following parameters are setup by the per-type "init" function.
```
Here is the added comment:

richsalz replied:
oops, the "/*" goes on a line by itself.

mspncp commented:
Ups, Murphys Law: forgot to rebuild and test...

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -127,7 +127,7 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
     unsigned char *nonce = NULL, *entropy = NULL;
     size_t noncelen = 0, entlen = 0;
```
`entroylen` would be more consistently named.

mspncp replied:
I already considered changing the name some time ago, because I hate inconsistent naming, too. But when I saw all the cryptic 'ent' names in drbgtest.c, I hesitated:
```
test/drbgtest.c:168:    t.entlen = td->entlen;
test/drbgtest.c:181:    t.entlen = td->entreseedlen;
test/drbgtest.c:199:    t.entlen = td->entlen_pr;
test/drbgtest.c:212:    t.entlen = td->entprlen_pr;
test/drbgtest.c:222:    t.entlen = td->entglen_pr;
```
So thanks for encouraging me to straighten out these names, too ;-)

paulidale replied:
Thanks.  That's even better.

mspncp commented:
Done.

* fixed comment nit
* replaced 'ent' by 'entropy'

To review my last changes only, use
```
git diff a4f7918317..0b23702d04
```

mspncp commented:
Done (second attempt). Sorry for generating so much build noise, I'm still practicing interactive rebasing :-|

richsalz commented:
merged as three commits :)  thanks!
